ironfish/src/network/peerNetwork.ts

```diff
@@ -657,24 +719,278 @@ export class PeerNetwork {
+        type: 'FULL',
+        value: new Transaction(nextFullTransaction.value.transaction),
+      })
+      nextFullTransaction = absoluteIndexTransactions.next()
+      fullTransactionsConsumed++
     }
+
+    return { ok: true, partialTransactions, missingTransactions }
   }
 
-  private handleNewBlockHashesMessage(peer: Peer, message: NewBlockHashesMessage) {
-    this.logger.debug(`Received unimplemented message ${message.type}`)
+  assembleBlockFromResponse(
+    partialTransactions: TransactionOrHash[],
+    responseTransactions: readonly SerializedTransaction[],
+  ): { ok: false } | { ok: true; transactions: Transaction[] } {
+    const transactions: Transaction[] = []
+    let currResponseIndex = 0
+
+    for (const partial of partialTransactions) {
+      if (partial.type === 'FULL') {
+        transactions.push(partial.value)
+      } else if (currResponseIndex >= responseTransactions.length) {
+        // did not respond with enough transactions
+        return { ok: false }
+      } else {
+        const next = new Transaction(responseTransactions[currResponseIndex])
+        if (!next.hash().equals(partial.value)) {
+          // hashes are mismatched
+          return { ok: false }
+        }
+        transactions.push(next)
+        currResponseIndex++
+      }
+    }
+
+    return { ok: true, transactions }
   }
 
-  private handleNewBlockV2Message(peer: Peer, message: NewBlockV2Message) {
-    this.logger.debug(`Received unimplemented message ${message.type}`)
+  private async onNewBlockTransactions(peer: Peer, message: GetBlockTransactionsResponse) {
+    const block = this.blockFetcher.receivedBlockTransactions(message)
+
+    if (!block) {
+      return
+    }
+
+    // if we don't have the previous block, start syncing
+    const prevHeader = await this.chain.getHeader(block.header.previousBlockHash)
+    if (prevHeader === null) {
+      this.chain.addOrphan(block.header)
+      this.node.syncer.startSync(peer)
+      return
+    }
+
+    await this.onNewFullBlock(peer, block, prevHeader)
+  }
+
+  private async onNewCompactBlock(peer: Peer, compactBlock: SerializedCompactBlock) {
+    if (!this.shouldProcessNewBlocks()) {
+      return
+    }
+
+    // mark the block as received in the block fetcher and decide whether to continue
+    // to validate this compact block or not
+    const shouldProcess = this.blockFetcher.receivedCompactBlock(compactBlock, peer)
+    if (!shouldProcess) {
+      return
+    }
+
+    // deserialize the header
+    const header = BlockHeaderSerde.deserialize(compactBlock.header)
+
+    // verify the header
+    const verifyHeaderResult = this.chain.verifier.verifyBlockHeader(header)
+    if (!verifyHeaderResult.valid) {
+      this.chain.addInvalid(header, verifyHeaderResult.reason ?? VerificationResultReason.ERROR)
+      this.blockFetcher.removeBlock(header.hash)
+      return
+    }
+
+    if ((await this.alreadyHaveBlock(header.hash)) || this.chain.isInvalid(header) !== null) {
```

Comment:
This `chain.isInvalid` sometimes adds blocks to invalid. But we also check if it's invalid in `alreadyHaveBlock`. I feel like we could split up the checking and the marking of invalid so it's more explicit here

Reply:
I'll see if I can refactor the isInvalid call to take either a header or a hash, and then propagate that through.

Reply:
Yea i was thinking it might make sense to not change the state of invalid in a function that is just called `isInvalid` but I think it is pretty safe here. We know that the input hashes to a certain header so it seems impossible that we'd accidentally mark something as invalid that is valid